### DISCUSSION TITLE
how to receive input from line in and output to headphones using the 'streams-audiokit-audiokit' example from arduino-audio-tools

### DISCUSSION THREAD

**Discussion by CodingArray** [pschatzmann/arduino-audio-tools#821] [Q&A]
Dear Phil Schatzmann,
I recently purchased an esp32 audio kit (AUDIOKIT_BOARD 5) type board and have been working on a project using your amazing library. I am extremely grateful to you every time I look at the source code.

I have tested most of the example programs, but there is one source code that doesn't work well, and I have a question about it.

I couldn't find an example of streaming the line input of the esp32 audio kit (AUDIOKIT_BOARD 5) type board to headphone output, and I found a similar example code. It works well from the built-in microphone input (AUDIO_HAL_ADC_INPUT_LINE2) to headphone output, but I'd like to know how to stream the line input to the headphone output.

#include "AudioTools.h"
#include "AudioLibs/AudioKit.h"

AudioKitStream kit; // Access I2S as stream
StreamCopy copier(kit, kit); // copy kit to kit

// Arduino Setup
void setup(void) {
    Serial.begin(115200);
    AudioLogger::instance().begin(Serial, AudioLogger::Warning);
    
    auto cfg = kit.defaultConfig(RXTX_MODE);
    cfg.sd_active = false;
    cfg.input_device = AUDIO_HAL_ADC_INPUT_LINE2;
    kit.begin(cfg);
}

// Arduino loop - copy data
void loop() {
    copier.copy();
}

**Comment by pschatzmann** [comment]
By default the output is sent to both output devices. So if you just want to send it to one  you would specify it in the config, like you did for the intput device with cfg.output_device (values are AUDIO_HAL_DAC_OUTPUT_LINE1, AUDIO_HAL_DAC_OUTPUT_LINE2,  AUDIO_HAL_DAC_OUTPUT_ALL)

Or you could just turn the amplifier off with kit.setSpeakerActive(false);

**Comment by CodingArray** [comment]
I have succeeded in streaming line input to headphone output using the following code that you advised me to try several times. I'm a little confused about the fact that 'AUDIO_HAL_ADC_INPUT_LINE2' is for the built-in microphone, but it worked when I set it up like below. Thank you once again.

#include "AudioTools.h"
#include "AudioLibs/AudioKit.h"

AudioKitStream kit; // Access I2S as stream
StreamCopy copier(kit, kit); // copy kit to kit

// Arduino Setup
void setup(void) {
    Serial.begin(115200);
    AudioLogger::instance().begin(Serial, AudioLogger::Warning);

    // Configure the Audio Kit
    auto cfg = kit.defaultConfig(RXTX_MODE);
    cfg.sd_active = false;

    // Use the line in input
    cfg.input_device = AUDIO_HAL_ADC_INPUT_LINE2; // Use LINE1 for line in input
    cfg.output_device = AUDIO_HAL_DAC_OUTPUT_LINE2; // Use LINE1 for headphone output
    kit.setVolume(100);
    // Start the Audio Kit
    kit.begin(cfg);
    kit.setSpeakerActive(false); // Turn off the speaker amplifier
}

// Arduino loop
void loop() {
    copier.copy();
}

**Reply by pschatzmann** [reply]
The AudioKit has a hardware bug: https://www.pschatzmann.ch/home/2021/12/15/the-ai-thinker-audiokit-audio-input-bug/

**Reply by CodingArray** [reply]
Your quick response saved me. Thank you so much. I want to be helpful to you. If I ever travel abroad, I would like to offer you a cold beer and coffee. I hope that your dedication will reduce the gap of information in this world...
